test/test_dataloader.py

```diff
@@ -769,10 +769,15 @@ def test_proper_exit(self):
 
                 try:
                     loader_p.join(JOIN_TIMEOUT + MP_STATUS_CHECK_INTERVAL)
-                    self.assertFalse(loader_p.is_alive(), desc + ': loader process not terminated')
+                    if loader_p.is_alive():
+                        fail_msg = desc + ': loader process did not terminate'
+                        if loader_p.exception is not None:
+                            self.fail(fail_msg + ', and had exception {}'.format(loader_p.exception))
+                        else:
+                            self.fail(fail_msg + ', and had no exception')
                     _, alive = psutil.wait_procs(worker_psutil_p, timeout=(MP_STATUS_CHECK_INTERVAL + JOIN_TIMEOUT))
```

Comment:
So... I know this is not part of this patch, but I have to ask. Is `worker_psutil_p` a list of PIDs? Don't we have a race condition here, where another process on the system could be allocated the same PID as the previous one, and then we hang indefinitely? (This shouldn't actually be the problem here, because you'd have to be allocating a lot of PIDs to wrap around)

Reply:
worker_psutil_p is a list of psutil.Process objects.  In _test_proper_exit, we ensure that the workers are alive when these objects are created. So they accurately track the status of worker processes.

Reply:
Specifically, this list is created at line 766 above, and then tester_setup_event is set. In _test_proper_exit (the function that the loader process is running), the loader process waits on tester_setup_event, and then makes sure that the workers are still is_alive() at lines 359-363.